### PR TITLE
Token trade plugin added

### PR DESCRIPTION
**Pull request by cbrzn**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -26,18 +26,18 @@ services:
       GRAPH_GRAPHQL_MAX_FIRST: '1000'
 
   ipfs:
-    image: daostack/test-env-experimental-ipfs:4.0.17
+    image: dorgtech/test-env-experimental-ipfs:4.0.21
```
Comment (orenyodfat) [inline docker-compose.yml:29]: please use daostack test env

File: package.json [hunk 2]
```diff
@@ -72,8 +72,8 @@
     ]
   },
   "devDependencies": {
-    "@daostack/migration-experimental": "0.1.2-rc.2-v0",
-    "@daostack/test-env-experimental": "4.0.17",
+    "@daostack/migration-experimental": "0.1.2-rc.4-v0",
+    "@dorgtech/test-env-experimental": "4.0.21",
```
Comment (orenyodfat) [inline package.json:76]: use daostack testenv

File: package.json [hunk 3]
```diff
@@ -72,8 +72,8 @@
     ]
   },
   "devDependencies": {
-    "@daostack/migration-experimental": "0.1.2-rc.2-v0",
-    "@daostack/test-env-experimental": "4.0.17",
+    "@daostack/migration-experimental": "0.1.2-rc.4-v0",
```
Comment (orenyodfat) [inline package.json:75]: 0.1.2-rc.6-v1

File: src/arc.ts [hunk 4]
```diff
@@ -627,6 +629,17 @@ export class Arc extends GraphNodeObserver {
     Logger.debug(`Data saved successfully as ${descriptionHash}`)
     return descriptionHash
   }
+
+  public approveTokens(tokenAddress: Address, spender: Address, amount: BN) {
```
Comment (orenyodfat) [inline src/arc.ts:633]: who is calling this function ?and on which cases ?

Comment (orenyodfat) [reply src/arc.ts:633]: please add test for that.

File: src/settings.ts [hunk 5]
```diff
@@ -1,6 +1,6 @@
 // the version of the redeemer contract to use.
 // we can specify multiple versions, and try to find them in the order given
-export const PACKAGE_VERSION = [0, 1, 2]
+export const PACKAGE_VERSION = [0, 1, 4]
```
Comment (orenyodfat) [inline src/settings.ts:3]: should it be 0,1,6 ?

File: src/token.ts [hunk 6]
```diff
@@ -13,6 +13,7 @@ import {
   ICommonQueryOptions,
   isAddress
 } from './index'
+import { Contract } from 'ethers'
```
Comment (orenyodfat) [inline src/token.ts:16]: is this needs here ?

File: test/utils.ts [hunk 7]
```diff
@@ -67,9 +67,9 @@ export interface ITestAddresses {
 }
 
 export function getTestAddresses(version: string = LATEST_ARC_VERSION): ITestAddresses {
-  const result = require('@daostack/test-env-experimental/daos.json').demo[version]
+  const result = require('@dorgtech/test-env-experimental/daos.json').demo[version]
```
Comment (orenyodfat) [inline test/utils.ts:70]: please use daostack .

File: src/arc.ts (hunk 4 above)
Comment (cbrzn) [reply src/arc.ts:633]: This is called on Alchemy when the user tries to create a proposal on TokenTrade Plugin. The use case is that the user needs to give approval to the TokenTrade Plugin instance of the DAO (before creating the proposal) so the tokens can be send to the DAO

File: src/arc.ts [hunk 8]
```diff
@@ -627,6 +629,18 @@ export class Arc extends GraphNodeObserver {
...
+    const erc20Abi = this.getABI({ abiName: 'ERC20Mock', version: LATEST_ARC_VERSION })
```
Comment (orenyodfat) [inline src/arc.ts:634]: I prefer not support/use ERC20Mock.
you can create the tokenContract just with the `approve` abi.(have it hard coded).

File: src/arc.ts [hunk 9]
```diff
@@ -41,6 +42,7 @@ import {
   Web3Client,
   Web3Provider
 } from './index'
+import { LATEST_ARC_VERSION } from './settings'
```
Comment (orenyodfat) [inline src/arc.ts:45]: this can be avoided .

File: test/utils.ts [hunk 10]
```diff
@@ -286,4 +286,4 @@ export function getContractAddressesFromMigration(environment: 'private' | 'rink
     }
   }
   return contracts
-}
+}
```
Comment (orenyodfat) [inline test/utils.ts:289]: redundent

File: src/arc.ts [hunk 11]
```diff
@@ -627,6 +628,31 @@ export class Arc extends GraphNodeObserver {
     Logger.debug(`Data saved successfully as ${descriptionHash}`)
     return descriptionHash
   }
+
+  public approveTokens(tokenAddress: Address, spender: Address, amount: BN) {
```
Comment (orenyodfat) [inline src/arc.ts:632]: any test which cover that ?

File: test/arc.spec.ts [hunk 12]
```diff
@@ -317,3 +317,13 @@ it('plugin contractInfo should contain alias', async () => {
...
 })
+
+it('arc.approveToken should return tx with status one', async () => {
+  const arc = await newArc()
+  const addresses = getTestAddresses()
+  const tokenAddress = addresses.dao.DAOToken
+  const account = '0x90f8bf6a479f320ead074411a4b0e7944ea8c9c1'
+  const approval = await arc.approveTokens(tokenAddress, account, new BN(1)).send();
+  expect(approval.transactionHash).toBeDefined();
+  expect(approval.receipt!.status).toEqual(1);
```
Comment (orenyodfat) [inline test/arc.spec.ts:328]: how about reading the token here and make sure there is an approval ?

**Review by orenyodfat — APPROVED** [review]

**Pull request merged**